Observable.from(iterable) generates Typescript compiler error

<!--
Thank you for raising your concerns, we appreciate your feedback and contributions to this repository.

Before you continue, consider the following:

If you have a "How do I do ...?" question, it is better for you and for us that this question is placed in [StackOverflow](http://stackoverflow.com/questions/tagged/rxjs5) or some chat channel. This way, you are making it easier for others to learn from your experiences too.

These "Issues" are meant only for technical problems, bugs, and proposals related to the library.

If your issue is a bug, please follow the format below:
-->

**RxJS version:** 5.0.3

**Code to reproduce:**
```
function* generateDoubles(seed: number) {
  let i = seed;
  while (true) {
    yield i;
    i = 2 * i; // double it
  }
}

const iterator = generateDoubles(3);
const result = Rx.Observable.from<number>(iterator).take(10);
result.subscribe((x: number) => console.log(x));
```

**Expected behavior:**
Since this is the example given on  [reactivex.io](http://reactivex.io/rxjs/class/es6/Observable.js~Observable.html#static-method-from)
It would be expected to compile in Typescript without an error.

**Actual behavior:**
```
error TS2345: Argument of type 'IterableIterator<number>' is not assignable to parameter of type 'ObservableInput<number>'.
  Type 'IterableIterator<number>' is not assignable to type 'PromiseLike<number>'.
    Property 'then' is missing in type 'IterableIterator<number>'.
```

**Additional information:**
This seems to be an issue with the definition of `ObservableInput<T>`:
```
export interface Subscribable<T> {
    subscribe(observerOrNext?: PartialObserver<T> | ((value: T) => void), error?: (error: any) => void, complete?: () => void): AnonymousSubscription;
}
export declare type SubscribableOrPromise<T> = Subscribable<T> | PromiseLike<T>;
export declare type ObservableInput<T> = SubscribableOrPromise<T> | ArrayLike<T>;
```

The iterable type is missing.

A fix might be to add:
```
export interface IterableLike<T> {
    [Symbol.iterator](): Iterator<T> | IterableIterator<T>;
}

// and fix ObservableInput<T>
export declare type ObservableInput<T> = SubscribableOrPromise<T> | ArrayLike<T> | IterableLike<T>;
```

**tsconfig.json**
```
{
	"compilerOptions": {
		"target": "es2015",
		"module": "commonjs",
		"moduleResolution": "node",
		"sourceMap": true,
		"alwaysStrict": true,
        "declaration": true,
		"noImplicitAny": true,
		"noImplicitThis": true,
		"strictNullChecks": true,
		"outDir": "dist",
		"lib": ["dom", "dom.iterable", "es2015", "es2015.iterable", "scripthost"]
	},
	"include": [
		"src"
	]
}
```

--- a/spec/observables/IteratorObservable-spec.ts
+++ b/spec/observables/IteratorObservable-spec.ts
@@ -29,7 +29,7 @@ describe('IteratorObservable', () => {
 
   it('should not accept boolean as iterator', () => {
     expect(() => {
-      IteratorObservable.create(false);
+      IteratorObservable.create(<any>false);
     }).to.throw(Error, 'object is not iterable');
   });
 
@@ -66,7 +66,7 @@ describe('IteratorObservable', () => {
 
     const results = [];
 
-    IteratorObservable.create(iterable)
+    IteratorObservable.create(<any>iterable)
       .take(3)
       .subscribe(
         x => results.push(x),
@@ -97,7 +97,7 @@ describe('IteratorObservable', () => {
 
     const results = [];
 
-    IteratorObservable.create(iterable, queue)
+    IteratorObservable.create(<any>iterable, queue)
       .take(3)
       .subscribe(
         x => results.push(x),
@@ -149,7 +149,7 @@ describe('IteratorObservable', () => {
     const expected = ['f', 'o', 'o'];
     IteratorObservable.create('foo')
       .subscribe(
-        (x: number) => { expect(x).to.equal(expected.shift()); },
+        (x) => { expect(x).to.equal(expected.shift()); },
         (x) => {
           done(new Error('should not be called'));
         }, () => {

--- a/src/Observable.ts
+++ b/src/Observable.ts
@@ -15,7 +15,8 @@ export interface Subscribable<T> {
 }
 
 export type SubscribableOrPromise<T> = Subscribable<T> | PromiseLike<T>;
-export type ObservableInput<T> = SubscribableOrPromise<T> | ArrayLike<T>;
+export type ArrayOrIterable<T> = Iterable<T> | ArrayLike<T>;
+export type ObservableInput<T> = SubscribableOrPromise<T> | ArrayOrIterable<T>;
 
 /**
  * A representation of any set of values over any amount of time. This the most basic building block

--- a/src/observable/FromObservable.ts
+++ b/src/observable/FromObservable.ts
@@ -1,5 +1,6 @@
 import { isArray } from '../util/isArray';
 import { isArrayLike } from '../util/isArrayLike';
+import { isIterable } from '../util/isIterable';
 import { isPromise } from '../util/isPromise';
 import { PromiseObservable } from './PromiseObservable';
 import { IteratorObservable } from'./IteratorObservable';
@@ -93,7 +94,7 @@ export class FromObservable<T> extends Observable<T> {
         return new ArrayObservable<T>(ish, scheduler);
       } else if (isPromise(ish)) {
         return new PromiseObservable<T>(ish, scheduler);
-      } else if (typeof ish[Symbol_iterator] === 'function' || typeof ish === 'string') {
+      } else if (isIterable(ish) || typeof ish === 'string') {
         return new IteratorObservable<T>(ish, scheduler);
       } else if (isArrayLike(ish)) {
         return new ArrayLikeObservable(ish, scheduler);

--- a/src/observable/IteratorObservable.ts
+++ b/src/observable/IteratorObservable.ts
@@ -1,9 +1,18 @@
 import { root } from '../util/root';
 import { IScheduler } from '../Scheduler';
-import { Observable } from '../Observable';
+import { Observable, ArrayOrIterable } from '../Observable';
 import { iterator as Symbol_iterator } from '../symbol/iterator';
 import { TeardownLogic } from '../Subscription';
 import { Subscriber } from '../Subscriber';
+import { isArrayLike } from '../util/isArrayLike';
+
+interface IteratorObservableState<T> {
+  subscriber: Subscriber<T>;
+  iterator: Iterator<T>;
+  index: number;
+  hasError?: boolean;
+  error?: any;
+}
 
 /**
  * We need this JSDoc comment for affecting ESDoc.
@@ -11,13 +20,13 @@ import { Subscriber } from '../Subscriber';
  * @hide true
  */
 export class IteratorObservable<T> extends Observable<T> {
-  private iterator: any;
+  private iterator: Iterator<T>;
 
-  static create<T>(iterator: any, scheduler?: IScheduler): IteratorObservable<T> {
-    return new IteratorObservable(iterator, scheduler);
+  static create<T>(iterator: ArrayOrIterable<T> | string, scheduler?: IScheduler): IteratorObservable<T> {
+    return new IteratorObservable<T>(iterator, scheduler);
   }
 
-  static dispatch(state: any) {
+  private static dispatch<T>(state: IteratorObservableState<T>) {
 
     const { index, hasError, iterator, subscriber } = state;
 
@@ -42,10 +51,10 @@ export class IteratorObservable<T> extends Observable<T> {
       return;
     }
 
-    (<any> this).schedule(state);
+    (<any>this).schedule(state);
   }
 
-  constructor(iterator: any, private scheduler?: IScheduler) {
+  constructor(iterator: ArrayOrIterable<T> | string, private scheduler?: IScheduler) {
     super();
 
     if (iterator == null) {
@@ -84,7 +93,7 @@ export class IteratorObservable<T> extends Observable<T> {
   }
 }
 
-class StringIterator {
+class StringIterator implements Iterator<string> {
   constructor(private str: string,
               private idx: number = 0,
               private len: number = str.length) {
@@ -92,41 +101,41 @@ class StringIterator {
   [Symbol_iterator]() { return (this); }
   next() {
     return this.idx < this.len ? {
-        done: false,
-        value: this.str.charAt(this.idx++)
+      done: false,
+      value: this.str.charAt(this.idx++)
     } : {
         done: true,
         value: undefined
-    };
+      };
   }
 }
 
-class ArrayIterator {
-  constructor(private arr: Array<any>,
+class ArrayIterator<T> implements Iterator<T> {
+  constructor(private arr: T[],
               private idx: number = 0,
               private len: number = toLength(arr)) {
   }
   [Symbol_iterator]() { return this; }
   next() {
     return this.idx < this.len ? {
-        done: false,
-        value: this.arr[this.idx++]
+      done: false,
+      value: this.arr[this.idx++]
     } : {
         done: true,
         value: undefined
-    };
+      };
   }
 }
 
-function getIterator(obj: any) {
+function getIterator<T>(obj: ArrayOrIterable<T> | string): Iterator<T> {
   const i = obj[Symbol_iterator];
-  if (!i && typeof obj === 'string') {
-    return new StringIterator(obj);
-  }
-  if (!i && obj.length !== undefined) {
-    return new ArrayIterator(obj);
-  }
   if (!i) {
+    if (typeof obj === 'string') {
+      return <any>new StringIterator(obj);
+    }
+    if (isArrayLike(obj)) {
+      return new ArrayIterator<T>(<any>obj);
+    }
     throw new TypeError('object is not iterable');
   }
   return obj[Symbol_iterator]();
@@ -137,17 +146,17 @@ const maxSafeInteger = Math.pow(2, 53) - 1;
 function toLength(o: any) {
   let len = +o.length;
   if (isNaN(len)) {
-      return 0;
+    return 0;
   }
   if (len === 0 || !numberIsFinite(len)) {
-      return len;
+    return len;
   }
   len = sign(len) * Math.floor(Math.abs(len));
   if (len <= 0) {
-      return 0;
+    return 0;
   }
   if (len > maxSafeInteger) {
-      return maxSafeInteger;
+    return maxSafeInteger;
   }
   return len;
 }

--- a/src/util/isIterable.ts
+++ b/src/util/isIterable.ts
@@ -1,0 +1,5 @@
+import { iterator as Symbol_iterator } from '../symbol/iterator';
+
+export function isIterable<T>(value: any): value is Iterable<T> {
+  return typeof value[Symbol_iterator] === 'function';
+}

